src/dmd/lexer.d

```diff
@@ -202,6 +203,104 @@ unittest
     }
 }
 
+
+/**************
+ * Pool allocator for Token objects, used by the Lexer.
+ * Tokens are never freed so Token pointers will never change
+ */
+private struct TokenPool
+{
+    private Array!Token tokens;
+
+    public ref Token getByIndex(size_t index) nothrow pure
+    {
+        if (index >= tokens.length)
+        {
+            tokens.setDim(index + 1);
+        }
+        return tokens[index];
+    }
+}
+
+
+/***************
+ * Represents a range of tokens lazily scanned from a lexer
+ * should be created with the Lexer makeRangeFromHere method
+ */
+struct TokenRange
+{
+private:
+    Lexer lex; //where are these tokens coming from?
+    size_t startIndex, stopIndex;
+
+public:
+    nothrow:
+    //range functions
+    Token front()
+    {
+        return lex.getByIndex(startIndex);
+    }
+
+    Token back()
+    {
+        return lex.getByIndex(stopIndex - 1);
+    }
+
+    void popFront() pure @nogc @safe
+    {
+        ++startIndex;
+    }
+
+    void popBack() pure @nogc @safe
+    {
+        assert(stopIndex > startIndex);
+        --stopIndex;
+    }
+
+    bool empty() const pure @nogc @safe
+    {
+        return startIndex == stopIndex;
+    }
+
+    size_t length() const pure @nogc @safe
```

Comment:
All these methods are undocumented. Even though they are standard range methods they won't show up in the documentation, so it's not possible to know that they're present without looking at the source code.

Reply:
Walter told me they didn't need ddoc comments since they're standard, should I just add /// above them?

Reply:
Yes, I noticed he mentioned that. I think you should add `///`, not sure if he will object.